### ISSUE TITLE
Allow Lanugage Navigation to be more effectively modified by custom extensions

### ISSUE THREAD

**Issue by HitmanInWis** [adobe/aem-core-wcm-components#503]
At Bounteous we were considering modifying the behavior of the Language Navigation to make one small changes.

Given this scenario:
- /content/site/en/lvl1/lvl2
- /content/site/fr/lvl1

WCM Core functionality:
- FR link on /content/site/en/lvl1.html goes to /content/site/fr/lvl1.html (same page)
- FR link on /content/site/en/lvl1/lvl2.html goes to /content/site/fr.html (home page)

Desired functionality:
- FR link on /content/site/en/lvl1.html goes to /content/site/fr/lvl1.html (same page)
- FR link on /content/site/en/lvl1/lvl2.html goes to /content/site/fr/lvl1.html (closest parent page)

Now I know we could submit a PR to WCM Core, and add a dialog option to change this behavior, but that would require us to then wait for the next WCM Core release.  Also, not every developer in the wild is looking to contribute back to WCM Core (for various reasons) - instead they want to *leverage* the tool.

The challenge with the current implementation and extension pattern is that to make this very slight change in component functionality a developer is basically forced to rewrite the entire sling model and dependency POJOs as custom code in the client codebase.

Before I discuss the challenges, let's discuss what this *could* look like if sling models were not hidden impl classes.  Note that this is not a discussion on the merits of hidden model impls and the proxy extension pattern, which we've debated on #466.  This is simply to show the difference of how simple this *could* be to implement versus how difficult it is to do so at current.

**Sunshine and bunnies solution**
If `LanguageNavigationImpl.class` was not hidden, custom development of a sling model extending this model could simply `@Override` the `getLocalizedPage()` function (it would need to be `protected` instead of `private`) to return the desired page path.  Everything else would remain as-is.   Clean, simple, done.

**Actual solution (given current constraints)**
With that out of the way, let's consider the "real" solution that can be achieved given the current patterns in WCM Core.

Since the developer is required to use the sling proxy extension pattern, the developer is required to  rewrite the logic to calculate the language links in their own implementation class.  In short, they must basically rewrite the entire `LanguageNavigationImpl` class *or* use the proxied super class to first `getItems()` and then change that `List<NavigationItem>` into a completely new list `NavigationItem` objects (gaining some minimal leverage at best).  Even if we go with that pattern, however, it now becomes even more challenging.

In overwriting `getItems()`, the developer is now forced to also overwrite the following line of code:
```
pages.add(new LanguageNavigationItemImpl(page, active, request, level, children, title));
```
However, `LanguageNavigationItemImpl` is also a hidden class, as is it's parent class `NavigationItemImpl`, as is it's parent class `PageListItemImpl`.  Because these are not sling models, there is no way to leverage the work WCM Core has done in these classes.  Rather, the developer must recode all of these classes into their own codebase in order to be able to build out the new `List<NavigationItem>` to be returned by their proxy extension of `LanguageNavigation`.

In the end, WCM Core requires the developer to rewrite `LanguageNavigationImpl`, `LanguageNavigationItemImpl`,  `NavigationItemImpl`, and `PageListItemImpl` in order to effect this simple change of logic in calculating the language link path.  Not only does the developer get no leverage from the WCM Core `languagenavigation` component, but the dev is also required to copy and paste the other `*ItemImpl` classes to their codebase with the possibility of them diverging over time from WCM Core where these classes are used also in other components.

**Now for potential solutions:**

Solution 1: Add `Page getLocalizedPage(Page page, Page languageRoot)` to the public interface of `LanguageNavigation`.  It feels a little convoluted, but would be a simple solution.  It would solve the issue of not being able to access `LanguageNavigationItemImpl`,  `NavigationItemImpl`, and `PageListItemImpl` at least for this particular scenario, because the developer would not need to overwrite `getItems()`.

Solution 2: Update `LanguageNavigationItemImpl`,  `NavigationItemImpl`, and `PageListItemImpl`, which are not sling models, to be public classes.  At least then when the proxy extension of `LanugageNavigationImpl` overwrites `getItems()` it can leverage these classes without overwriting them.  Even if we go with Solution 1, it might be worth considering this as well since its only a matter of time before another scenario comes up that needs access to these classes.

Solution 3: My prefered solution (as discussed in #466) is still to make all WCM Core model classes exported so they can be directly extended.  It doesn't seem many people agree with me on that though :).

Thoughts?

**Comment by vladbailescu** [comment]
As you know, we always prefer solution 0 (PR to Core WCM Components, with a content policy option to control behaviour for this case) as it will benefit everyone else using the components.

My objections to your proposed solutions:
* Solution 1: We try to only export properties that can be used from HTL or methods that are used for JSON serialisation in our model interfaces.
* Solution 2 & 3: As you note, making just the implementation classes public is not always enough, we would need to also make other aspects of the implementation public/protected. This, in time, leads to a very large API surface that we will need to make sure not to break (and we already proved we're kind of lousy at that 😄)

There's also solution 4: you copy the implementation classes into your project and hack away.

**Comment by HitmanInWis** [comment]
Agreed that solutions #1 and #2 aren't great.  Solution #4 is what I mentioned above as the current only solution, but removes any leverage I get as an engineer using WCM Core and potentially leaves me in a state where some of my language nav links work one way and others work a different way b/c my component uses my versions of `LanguageNavigationItemImpl`, `NavigationItemImpl`, and `PageListItemImpl` which may become out of date with versions in WCM Core that are used by other components.

If we want the community to adopt WCM Core components it's the engineers that are going to lead the way.  It's not realistic to think otherwise.  And if we want engineers to use the tool, it's going to have to provide compelling benefits and ease versus not using the tool.  Even if it's a "good idea" engineers are going to do what they are used to and/or what's expedient, and I feel roadblocks like what I've described in this ticket are going to slow down adoption of WCM Core as people revert to "easier" methods.

I understand that Solution #0 (a PR to WCM core) is preferred, but like I said, though Bounteous is very active in the Adobe open source community and happy to create PRs, this requires us to wait for the next release to be able to use, and frankly not everyone in the community is willing to participate at that level.  Regardless of the community's willingness to contribute to the tool, I believe the goal of this library is to increase adoption to nearly 100%.  To do so, we need to make it more appealing to engineers.  I'm just one voice (because we didnt want to pile a bunch of Bounteous devs on here saying "I agree with Brett") but I do speak for more than just myself on this matter.

**Comment by msullivan-r2i** [comment]
I agree 100% with Brett.

These shouldn't be thought of as a product that you have to support and not break the API, but building blocks for developers.

I would argue that having to duplicate any classes adds more technical debt to a project compared to running the risk of the API breaking.  You would have to re-evaluate all of these duplicated classes for each release, as opposed to incrementing the core components version number and testing.

**Comment by vladbailescu** [comment]
Hi @msullivan-r2i and thanks for joining the conversation!

I totally agree that our focus should be to help developers. We considered providing a concise and stable API as one of the methods to do this. And if we did it right this should even allow (some sort of) automated updates.

Of course, there are many other things we could improve based on feedback from developers (like yours) which we do discuss internally seriously. But changing this pattern is a bigger discussion and we will need to consider all aspects and discuss with many more teams.

I see the biggest reason cited for not contributing a change as frequency of releases. There's actually an imminent 2.3.2 release and we are already discussing a 2.4.0 (which would also include the fix for #491). We took some time and smoothed out our internal validation and release process so that we can do more often releases (ideally we would do at least a release every month). We're trying to remove this blocker.

**Comment by vladbailescu** [comment]
I updated the title to better reflect what we're discussing (the Language Navigation is just an example).  Hope that's ok, @HitmanInWis

**Comment by HitmanInWis** [comment]
@vladbailescu for sure, thanks for updating the title.

Can you dig further into the idea of trying to support automated updates of WCM Core?  I'm not seeing obvious benefit of this since:
- Anyone using WCM Core components is likely to at least be layering on styling, so any updates need to be tested unless they are versioned
- If changes are indeed versioned, then an auto-upgrade really doesnt give the end user/engineer anything either, since component extensions still reference the old version.

I know it's been discussed in #466 that there is a desire to support updating WCM Core components on a server w/o re-deploying the client code, but I struggle to see the real application where someone would actually want to do that (and also accept any associated risks in doing so, given it relies on perfect backward compatibility and involves production systems).

**Comment by vladbailescu** [comment]
@HitmanInWis , one the common scenarios we've seen with our customers is having a website that was developed in the past with a team that's no longer available/convenient to work with. When they want to build a new website on the same infrastructure they might want to use newer features/components. However, since they are shared across sites, they are unwilling to upgrade due to fear of breaking the existing sites.

Another scenario is where an ops team wants to keep the platform always up-to-date (for security or performance reasons). They could have a staging environment for validating updates but ideally they should be able to proceed without always going back to the development team.

As you notice, there's a very fine balance between pushing updates/features and keeping backwards compatible. We're still learning how to manage this.

We do test internally some upgrades scenarios. They are mostly verifying We.Retail and a minimal project (created with the https://github.com/adobe/aem-project-archetype) _survive_ an update. If you have ideas about improving these tests, do let us know.

**Comment by HitmanInWis** [comment]
To be clear, I'm not advocating for non-backward compatibility.  I think versioning and backward compatibility are great things.

Versioning and backward compatibility are the things needed to satisfy your scenario of a client that no longer has their previous implementation partner to be able to use the old components on the old code, yet the newer components on the newer code.  The proxy inheritance pattern with hidden implementation classes is not what enables that.

The proxy inheritance pattern *may* be important for the second scenario - the ability to update WCM Core for a security fix w/o rebuilding the client code.  Do we 100% know if descendant classes in the client code would need to be rebuilt for the new WCM Core classes if the overall interface of the WCM Core class is backward compatible?

**Comment by vladbailescu** [comment]
Agree that proper versioning policies and enforcement are what save us from breaking implementations.

The proxy pattern allows us to freely refactor the implementations. Exporting the implementations and enlarging the number of methods that are public/protected (such as `getLocalizedPage()` for your examples) would be restricting this freedom (as they now become API).

**Comment by HitmanInWis** [comment]
I would posit, though, that the goal of a prolific, reusable, shared library should shift any pain to the repo maintainers (by *slightly* restricting freedom of changing implementation) rather than putting that pain on the end users of the library.  I also again question how much "actual" over "theoretical" pain this would result in.

Curious what others' thoughts are beyond myself and @vladbailescu.  The more different opinions, the better.

**Comment by stevengssns** [comment]
In #490 I also made the remark that the core components are lacking extensibility. What is being discussed here seems to come very close to what I've experienced so far.

So my 2 cents:


First of all, I fully support the [delegation pattern](https://github.com/adobe/aem-core-wcm-components/wiki/Delegation-Pattern-for-Sling-Models). 

This perfectly hides any implementation details, and creates a clear contract between the framework developers and their clients (the sling model exporter can still leak implementation details by exporting implementation specific properties, but this can be considered as a fixable coding bug).

If a team is happy to depend on implementation details, they could just as well fork the whole project, and roll their own.


I also understand the framework maintainer's desire to keep the API surface limited, as this will greatly improve maintainability. 
That being said, I do have the feeling that not enough thought goes into designing for extensibility. 

The example that started this issue is a good one. It doesn't take a lot of analysis to come to the conclusion that there are little or no extension possibilities for the LanguageNavigation model. As described, there is only the getItems() method that can be manipulated, which will quickly result in a full reimplementation (possibly rippling to other models).

When I look at proposed solution #1 "Add Page getLocalizedPage(Page page, Page languageRoot) to the public interface of LanguageNavigation", then my first reaction is.. what is this logic doing on LanguageNavigationImpl in the first place? Surely this should be a "simple" getter on the page (sling) model (if we are taking OO seriously)? There it should not feel funny to make it public.
If LanguageNavigationImpl would then leverage the page (sling) model to get it's localized variant, then this should solve this specific extension issue, because the getLocalizedPage() getter could be changed through the delegation pattern.
I do know this might be more easily said than done.... And furthermore, the getItems() documentation should explicitly explain that this getLocalizedPage() getter will be called, so that it becomes part of the contract (so obviously there should be a test aswel).


Another example of "lack of attention to extensibility" is the massive use of @PostConstruct. Most, if not all, of the code that is placed there will be out of reach of the clients. E.g. a lot of variables are initialised here, often with rules for deriving default values. In several cases it might be convenient to manipulate them to drive the component's behaviour in a way that is desirable.
Simply making sure that these getters are made public, and ensuring that the rest of the implementation code uses the getter instead of the variable, will make this possible. As a bonus it also becomes possible to add lazy initialisation to defer code execution (only if there might be code paths that will never call the getter of course.).
For the LanguageNavigation model I wouldn't mind having a public getNavigationRoot() and getStructureDepth() getter so that I could change the way they are obtained. 


Another example I can give is the Teaser component, where it is not possible to insert additional HTML like for instance a subtitle. This could be fixed very easily by simply adding empty placeholder templates/scripts into the main sightly script.
Maybe I am being naive, but simply asking the question "What might a developer want to customize?" should quickly yield this as one of the possibilities.


I will probably be able to find other examples, but they will probably be more case specific. 

 
So in conclusion, I understand that there are (and should be) limits to the extensibility, but I also believe too little thought and effort are being invested in designing for extensibility.

**Comment by klcodanr** [comment]
For what it's worth, I would agree that at the moment, the design of the core components inhibits extensibility to an unacceptable level. The old foundation components were bad because of poor coding practices, but they did offer an advantage in that they could be read and extended easily, the core components, in my opinion, have swung too far in the other direction. 

The problem, as I see it, is that a lot of the logic that by all rights, should be display or business logic,  is instead locked into implementation classes which cannot be extended, but instead, need to be replaced wholesale. 

For a practical example, look at the ImageImpl:
https://github.com/adobe/aem-core-wcm-components/blob/master/bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v2/ImageImpl.java

What if your implementation had different logic around how the description was sourced? For example, It used a custom field? This would require re-writing the whole class instead of just overriding that single field which puts implementors on a MUCH worse upgrade trajectory than simply overriding a single class. 

Additionally, is there a reason that this logic could not be configurable? Heck, why not require us developers to just deploy CA Configs along with our apps if we want to change the behavior? I know that for whatever unknowable reason Adobe has not adopted CA Configs, but even as a developer configuration tool, this would be 100x better than having to override a whole class to change what attribute gets used for the description of an image. 

The question that needs to be asked here is what is the purpose of the core components? Are they meant to be a set of building blocks that can be extended and customized or are they a rigid example of how to implement components? Right now they're the latter, but I'd argue it'd be much more useful for the community if they were the former.

**Comment by joerghoh** [comment]
From what I understand, there are parties involved in here with different goals:

* Adobe wanting to make something with a limited API surface, which can be updated without having to much risk. This means that Adobe wants to limit the API surface .
* Developers want to extend/modify the components provided by Adobe, and they need to augment/change the existing implementation without the need to duplicate existing code (frontend and/or backend).

I accept both as valid points, but at the moment the developers are a bit limited (as pointed out) in the way how to (re-) use the core components. Changing a single property name often results in copy&paste of a complete component, and that is really hard to justify.

I will try to outline some requirements I would think are critical:
* A developer should be able to replace the Sling Model provided by the Core Components with his own version, without the need to modify the HTL (and all related frontend stuff). This mandates that there is a contract (an API) on what HTL can rely on. At that point the implementation doesn't matter.
* There should be a default implementation. This default implementation is part of the Core Components and it is built in a way, that a developer can inherit from this code and implement his own logic.
* And somehow (without the need to change the HTL) the developer causes HTL to use his own model implementation instead of the default model of Core Components. Maybe OSGI service ranking? I don't know if that's possible already.

This approach would allow a developer to reuse the Core Component model or replace it by his own version. Versioning is possible by the means of standard OSGI mechanics.

I don't think that this approach is different from the classical "use the library" vs "extend the library" questions we have since decades. In OSGI we already have solutions for it, and maybe we can study their ideas and adapt them to the exact usecases we have here. 
But in the end it's a question of adoption, and in order to make this Core Components library being used by all projects we should cover most (if not all) usecases. 

And for the aspect of backwards-compatibility: we should design our APIs carefully, so we don't need to break them every other month :-)

just some Cents by a random non-developer

**Comment by kwin** [comment]
IMHO changing the model without touching the HTL by implementing a model with a higher ranking is an Anti-Pattern as that modifies the behavior of this component globally (i.e. you potentially destroy other reference sites like we-retail)

**Comment by joerghoh** [comment]
@kwin maybe we need a proxy layer then, which binds the correct model classed based on path and whatsoever.

**Comment by HitmanInWis** [comment]
Unless I'm mistaken, the issue of "And somehow (without the need to change the HTL) the developer causes HTL to use his own model implementation instead of the default model of Core Components." is already addressed when you make a proxy component extension in your own codebase.

Since the WCM Core HTL references sling model interfaces, as long as your sling model impl is linked to your proxy component extension resource type and extends the same interface as WCM Core, then the HTL from WCM Core does indeed pick up your custom sling model implementation class.  The problem is, as @stevengssns and @klcodanr  have pointed out, that sling model is currently very limited in how it can override small portions of the WCM Core model rather than completely replacing it.

Or am I misunderstanding the ask?

**Comment by joerghoh** [comment]
@HitmanInWis I guess you are right.

I think we should come up with a detailled of usecases how you can extend Core Components (maybe in a separate documentation area of this repo), and list them all here. For example:

* How do I change the markup?
* How do I switch to my own model
* I want to use Core Components models, but tune some pieces of it.
* etc.

This would get us away from this lengthy discussion, and forces us to focus on exact usecases and how to address them.

And next to it it gives a great piece of documentation :-)

**Comment by stefanseifert** [comment]
i started a likewise discussion some time ago in #65 - i think we have to differentiate between the different types of core components. some are very simple like title, text etc. and no problem arises as there is no much need to access/customize the logic hidden in the model.

the main problem is were too much logic is packed into the model of a single component (like the image component) which should rather be an "aspect" to be used by all models, and not hidden into one single model. this are thinks like:
* image/media handling - displaying images, resizing, cropping, rendition selection, image markup etc.
* link handling - resolving and externalizing links
* site root detection

all this should not at all be part of a model, but part of separate API classes with their own versioning and used by the models. this would dramatically reduce the locic in some of the core components and reduce the need for overlaying only parts of the model logic which is always difficult.

btw. modeling such thinks as media, link, root detection handling in a separate API to be reusable in all component over the whole projects is what we're doing for years in http://wcm.io/handler/

this was always an important part of functionality missing in AEM itself, the old foundation components and also the current core components.

the current pattern of re-using image functionality in other components by embedding the or extending the image component is very inflexible, and a bad pattern imho (as it is done e.g. in the teaser component).

**Comment by rajeevyadav2** [comment]
@HitmanInWis : We might also look into ClientLibs and HTL extensibility of these components. Currently,  JS is too tightly coupled with component functionality. If someone wants to customize the JS code to extend the core component functionality then they need create a copy of current JS and make the changes. Now with this done, if core components comes up with any new code changes in JS for bug fixes/functionality extension then customized component does not get it as they have overridden the JS and pointing to local version of it. 
It would be good to have some JS pattern implemented in clientlib which allows consumers to extend the JS code but at the same time if they move to newer core version, they should be able to get core fixes/updates.
